src/careamics/dataset/iterable_dataset.py

```diff
@@ -125,36 +131,72 @@ def _calculate_mean_and_std(self) -> StatsOutput:
         PatchedOutput
             Data class containing the image statistics.
         """
-        image_means = []
-        image_stds = []
-        target_means = []
-        target_stds = []
         num_samples = 0
 
         for sample, target in iterate_over_files(
             self.data_config, self.data_files, self.target_files, self.read_source_func
         ):
-            sample_mean, sample_std = compute_normalization_stats(sample)
-            image_means.append(sample_mean)
-            image_stds.append(sample_std)
+            # compute mean and std for each sample
+            init_mean, _ = compute_normalization_stats(sample)
+            # separate channels
+            sample_channels = np.array(np.split(sample, sample.shape[1], axis=1))
+
+            if num_samples == 0:
+                mean = init_mean
+                count = np.array(
+                    [np.prod(channel.shape) for channel in sample_channels]
+                )
+                m2 = np.array(
+                    [
+                        np.sum(
+                            np.subtract(channel.flatten(), [mean[i]] * count[i]) ** 2
+                        )
+                        for i, channel in enumerate(sample_channels)
+                    ]
+                )
+            else:
+                count, mean, m2 = update_iterative_stats(
+                    count, mean, m2, sample_channels
+                )
 
             if target is not None:
-                target_mean, target_std = compute_normalization_stats(target)
-                target_means.append(target_mean)
-                target_stds.append(target_std)
+                target_init_mean, _ = compute_normalization_stats(target)
+                target_channels = np.array(np.split(target, target.shape[1], axis=1))
+                if num_samples == 0:
```

Comment:
This should be refactored somewhere else, especially because the whole procedure is called twice!

In the spirit of the previous comment, `update_iterative_stats` and the other method could become `_update_iterative_stats` and then a new method in `dataset_utils.welford` called `welford_mean_and_std(array, ...)` could do the calculation